sources/fetchUtils.ts

```diff
@@ -0,0 +1,143 @@
+import {UsageError} from 'clipanion';
+import fs           from 'node:fs';
+import path         from 'node:path';
+
+let mocks: Map<string, {
+  body: ArrayBuffer;
+  status: number;
+  headers: Record<string, string>;
+}>;
+
+if (process.env.NODE_ENV === `test` && process.env.NOCK_ENV === `record`) {
+  process.once(`exit`, () => {
+    if (!mocks)
+      return;
+
+    fs.mkdirSync(path.dirname(getNockFile()), {recursive: true});
+    fs.writeFileSync(getNockFile(), JSON.stringify(Array.from(mocks.entries()), (key, value) => {
+      return value instanceof ArrayBuffer ? Buffer.from(value).toString(`base64`) : value;
+    }, `\t`));
+  });
+}
+
+function getMocks() {
+  if (mocks)
+    return mocks;
+
+  if (!fs.existsSync(getNockFile()))
+    throw new Error(`No nock file found for this test run; run the tests with NOCK_ENV=record to generate one`);
+
+  mocks = new Map(JSON.parse(fs.readFileSync(getNockFile(), `utf8`), (key, value) => {
+    return typeof value === `string` && key === `body` ? Buffer.from(value, `base64`) : value;
+  }));
+
+  return mocks;
+}
+
+export async function fetch(input: string | URL, init?: RequestInit) {
+  if (process.env.COREPACK_ENABLE_NETWORK === `0`)
+    throw new UsageError(`Network access disabled by the environment; can't reach ${input}`);
+
+  if (process.env.NODE_ENV !== `test`) {
+    return fetchWrapper(input, init);
+  } else if (process.env.NOCK_ENV === `record`) {
+    const response = await fetchWrapper(input, init);
+    const data = await response.arrayBuffer();
+
+    const minimalHeaders = new Headers();
+    if (response.headers.has(`content-type`))
+      minimalHeaders.set(`content-type`, response.headers.get(`content-type`)!);
+    if (response.headers.has(`content-length`))
+      minimalHeaders.set(`content-length`, response.headers.get(`content-length`)!);
+
+    mocks ??= new Map();
+    mocks.set(input.toString(), {
+      body: data,
+      status: response.status,
+      headers: Object.fromEntries(minimalHeaders),
+    });
+
+    return new Response(data, {
+      status: response.status,
+      headers: minimalHeaders,
+    });
+  } else if (process.env.NOCK_ENV === `replay`) {
+    const mocks = getMocks();
+
+    const mock = mocks.get(input.toString());
+    if (!mock)
+      throw new Error(`No mock found for ${input}`);
+
+    return new Response(mock.body, {
+      status: mock.status,
+      headers: mock.headers,
+    });
+  } else {
+    return fetchWrapper(input, init);
+  }
+}
+
+export async function fetchJSON(input: string | URL, init?: RequestInit) {
+  const response = await fetch(input, init);
+  const text = await response.text();
```

Comment:
```suggestion
  const text = await response.json();
```

Reply:
That defeats the purpose of this function; to provide a custom error if `JSON.parse` fails.
It's a port of https://github.com/nodejs/corepack/blob/6df5063b14868ff21499a051e5122fa7211be6ed/sources/httpUtils.ts#L58-L71

Reply:
Here's the error message returned by Undici: `SyntaxError: Unexpected token '<', "<html><hea"... is not valid JSON`. IMO it's good enough, we don't gain much by rewriting the logic ourselves. No strong feelings though.

Reply:
That seems to be only in Node.js v19 and up, tested with
```js
Promise.resolve()
	.then(() => {
		JSON.parse('<html>');
	})
	.catch(console.error);
```
In Node.js v18 and under the error is:
```
SyntaxError: Unexpected token < in JSON at position 0
```

Reply:
The error stack trace is prefixed by the whole string though:

```console
$ node --input-type=module -e '(await fetch("data:text/html,<html>whatever</html>")).json()'
undefined:1
<html>whatever</html>
^

SyntaxError: Unexpected token < in JSON at position 0
    at JSON.parse (<anonymous>)
    at parseJSONFromBytes (node:internal/deps/undici/undici:4747:19)
    at successSteps (node:internal/deps/undici/undici:4718:27)
    at fullyReadBody (node:internal/deps/undici/undici:1433:9)
    at async specConsumeBody (node:internal/deps/undici/undici:4727:7)

Node.js v18.19.0
```

Reply:
Its even better when you use a file instead of eval because it shows the file and line number:

```
$ node fetch-error.mjs
undefined:1
<html>whatever</html>
^

SyntaxError: Unexpected token < in JSON at position 0
    at JSON.parse (<anonymous>)
    at parseJSONFromBytes (node:internal/deps/undici/undici:4747:19)
    at successSteps (node:internal/deps/undici/undici:4718:27)
    at fullyReadBody (node:internal/deps/undici/undici:1433:9)
    at async specConsumeBody (node:internal/deps/undici/undici:4727:7)
    at async file:///Users/styfle/Code/example/fetch-error.mjs:3:14

Node.js v18.19.0
```

Reply:
Alright, removed the function, the error wont be the same between the different Node.js versions but I suppose that's acceptable.